File: number-of-1-bits/solution191.py
# https://leetcode.com/problems/number-of-1-bits/

class Solution191:

    # T: O(1), S: O(1)
    def hammingWeight(self, n: int) -> int:
        count, mask = 0, 1
        for _ in range(32):
            if n & mask:
                count += 1
            n >>= 1
        return count

    # T: O(1), S: O(1)
    def hammingWeightMaskShifting(self, n: int) -> int:
        count, mask = 0, 1
        for _ in range(32):
            if n & mask:
                count += 1
            mask <<= 1
        return count


def main():
    sol = Solution191()
    n1 = 0b00000000000000000000000000001011
    n2 = 0b00000000000000000000000010000000
    print(sol.hammingWeight(n1))  # 3
    print(sol.hammingWeightMaskShifting(n1))  # 3
    print(sol.hammingWeight(n2))  # 1
    print(sol.hammingWeightMaskShifting(n2))  # 1


if __name__ == "__main__":
    main()

File: number-of-1-bits/Solution191.java
// https://leetcode.com/problems/number-of-1-bits/

class Solution191 {

    // T: O(1), S: O(1)
    public int hammingWeight(int n) {
        int count = 0, mask = 1;
        for (int i = 0; i < 32; i++) {
            if ((n & mask) == 1) {
                count++;
            }
            n >>= 1;
        }
        return count;
    }

    // T: O(1), S: O(1)
    public int hammingWeightMaskShifting(int n) {
        int count = 0, mask = 1;
        for (int i = 0; i < 32; i++) {
            if ((n & mask) != 0) {
                count++;
            }
            mask <<= 1;
        }
        return count;
    }

    public static void main(String[] args) {
        var sol = new Solution191();
        int n1 = 00000000000000000000000000001011;
        int n2 = 00000000000000000000000010000000;
        System.out.println(sol.hammingWeight(n1)); // 3
        System.out.println(sol.hammingWeightMaskShifting(n1)); // 3
        System.out.println(sol.hammingWeight(n2)); // 1
        System.out.println(sol.hammingWeightMaskShifting(n2)); // 1
    }

}
